spec/02-integration/000-new-dao/03-db_cluster_mutex_spec.lua

```diff
@@ -0,0 +1,112 @@
+local helpers = require "spec.helpers"
+
+
+for _, strategy in helpers.each_strategy("cassandra") do
+  describe("kong.db [#" .. strategy .. "]", function()
+    local db
+
+
+    setup(function()
+      local _
+      _, db, _ = helpers.get_db_utils(strategy)
+    end)
+
+
+    describe("db:cluster_mutex()", function()
+      it("returns 'true' when mutex ran and 'false' otherwise", function()
+        local t1 = ngx.thread.spawn(function()
+          local ok, err = db:cluster_mutex("my_key", nil, function()
+            ngx.sleep(0.1)
+          end)
+          assert.is_nil(err)
+          assert.equal(true, ok)
+        end)
+
+        local t2 = ngx.thread.spawn(function()
+          local ok, err = db:cluster_mutex("my_key", nil, function() end)
+          assert.is_nil(err)
+          assert.equal(false, ok)
+        end)
+
+        ngx.thread.wait(t1)
+        ngx.thread.wait(t2)
+      end)
+
+
+      it("mutex ensures only one callback gets called", function()
+        local cb1 = spy.new(function() end)
+        local cb2 = spy.new(function() ngx.sleep(0.2) end)
+
+        local t1 = ngx.thread.spawn(function()
+          ngx.sleep(0.2)
+
+          local _, err = db:cluster_mutex("my_key_2", { owner = "1" }, cb1)
+          assert.is_nil(err)
+        end)
+
+        local t2 = ngx.thread.spawn(function()
+          local _, err = db:cluster_mutex("my_key_2", { owner = "2" }, cb2)
+          assert.is_nil(err)
+        end)
+
+        ngx.thread.wait(t1)
+        ngx.thread.wait(t2)
+
+        assert.spy(cb2).was_called()
+        assert.spy(cb1).was_not_called()
```

Comment:
since cb2 takes 0.2 to run and t1 waits 0.2 until it acquires, isn't there a chance we get unlucky in this test and t2 releases the lock before t1 tries to acquire? (in other words, shouldn't the sleep in cb2 be a little longer than t1 to be on the safe side, or am I misunderstanding the test?)